packages/itwinui-react/src/core/Table/Table.tsx

```diff
@@ -816,8 +823,41 @@ export const Table = <
     }
   });
 
+  const virtualizer = useVirtualizer({
+    count: page.length,
+    getScrollElement: () => tableElement ?? null,
+    estimateSize: () => 62, //Set to 62px since that is the default height of a table row.
+    overscan: 10,
+    indexAttribute: 'data-iui-index',
+  });
+
+  const outerProps = {
+    style: {
+      minBlockSize: virtualizer.getTotalSize(),
+      minInlineSize: '100%',
+      ...style,
+    },
+    ...rest,
+  } as React.HTMLAttributes<HTMLElement>;
+
+  const innerProps = {
+    style: { willChange: 'transform' },
```

Comment:
I believe the `<div {...innerProps}>` is no longer needed anywhere. The new virtualizer only requires a single wrapping div instead of two.

Reply:
Removed the `innerProps` object and `<div>` element in this [commit](https://github.com/iTwin/iTwinUI/pull/2092/commits/105098f1819da3cfb8b47a090754d51ccf22bf51). For the `Tree` component, the `innerProps` are going on an `<ul>` inside of the outer `<div>` component. On top of that, trying to put the styles from the `innerProps` onto this outer `<div>` causes the entire tree to take the height of the `virtualizer.getTotalSize()` value, which defeats the point of the virtualization. Is it fine to leave the `Tree` component with its `innerProps` for now?

Reply:
I'm not sure I understand the problem. Didn't we talk about not having an outer `<div>` in `Tree`? Can you share the code you're trying to use that's causing the problem?

Reply:
There was the outer `<div>` in the react workshop example that was outside of the `Tree` element itself, which is what made the recursive search for a scrollable parent necessary. That was removed, but there is another outer `<div>` inside of the `Tree` component that is now acting as the scrollable parent for the virtualization. Putting the full size on the scrollable parent breaks the virtualization, since the virtualizer tries to load all of the items at the same time. This isn't an issue for the `Table` and `ComboBox` component, since they have higher level scrollable elements that the scroll ref can point to instead. `Tree` doesn't have an element like that, though, so it has to wrap the `TreeElement` in a `<div>` that is scrollable. There can't be an inner `<div>` with the height while the `TreeElement` is the scrollable element because the `TreeNode` elements are `<li>` elements at their top level. Therefore, it seems like the `innerProps` and `outerProps` structure is still needed on `Tree`, unless I am completely missing something. Broken code that loads in every element at once:

```jsx
 const outerProps = {
      className: className,
      style: {
        minInlineSize: '100%',
        overflow: 'auto',
        minBlockSize: virtualizer.getTotalSize(),
        willChange: 'transform',
        ...style,
      },
    } as React.HTMLAttributes<HTMLElement>;

    const innerProps = {
      style: {},
    } as React.HTMLAttributes<HTMLElement>;

    return (
      <Box {...outerProps} ref={parentRef}>
        <TreeElement {...innerProps} {...rest} ref={ref}>
          {virtualizer
            .getVirtualItems()
            .map((virtualItem) =>
              itemRenderer(virtualItem.index, virtualItem, virtualizer),
            )}
        </TreeElement>
      </Box>
    );
```

Reply:
I don't think we should be adding a wrapper div around `TreeElement`. The `...rest` props and forwarded `ref` should almost always be spread on the topmost element of the component.

But we should be able to add an extra div *inside* `TreeElement`, and even hide it inside shadow DOM, similar to https://github.com/iTwin/iTwinUI/pull/2092#discussion_r1644627752. Can you explain why this is a problem: "the `TreeNode` elements are `<li>` elements at their top level"?

Reply:
So the `TreeElement` component is a polymorphic `<ul>`, and the `TreeNode` component returns a `<li>` element, so, from my understanding, putting a `<div>` between the two elements would create an invalid DOM, since the `<li>` needs a list element as its parent, and the `<div>` would be the parent in that case. Wrapping the inner `<div>` in the shadow DOM would make the light DOM valid, though, so I've been trying to implement that. However, I am currently getting these errors when I try to attach a `ShadowRoot` to the `TreeElement`:
![image](https://github.com/iTwin/iTwinUI/assets/141063311/bf6d6878-e26a-41e8-af6a-b8ce59226cde)
I am not sure what the cause of this is. Could it be because the `TreeElement` is being defined by `polymorphic`? code:
```jsx
const TreeElement = polymorphic.ul('iui-tree', {
  role: 'tree',
  tabIndex: 0,
});


return (
      <TreeElement
        {...outerProps}
        {...rest}
        ref={useMergedRefs(ref, parentRef)}
      >
        <ShadowRoot>
          <div {...innerProps}>
            <slot />
          </div>
        </ShadowRoot>
        <>
          {virtualizer
            .getVirtualItems()
            .map((virtualItem) =>
              itemRenderer(virtualItem.index, virtualItem, virtualizer),
            )}
        </>
      </TreeElement>
```

Reply:
That's a very valid point about invalid DOM. We can change the `<ul>` to be `<div>` though. The `<ul>`+`<li>` doesn't give us anything useful, because we are manually setting their respective roles to [`"tree"`](https://github.com/iTwin/iTwinUI/blob/3b5fe97caf19d4314e0f91645c9ec6b9cdb1c8d8/packages/itwinui-react/src/core/Tree/Tree.tsx#L358) (overriding [implicit `"list"`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/ul#technical_summary)) and [`"treeitem"`](https://github.com/iTwin/iTwinUI/blob/3b5fe97caf19d4314e0f91645c9ec6b9cdb1c8d8/packages/itwinui-react/src/core/Tree/TreeNode.tsx#L256) (overriding [implicit `"listitem"`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/li#technical_summary)). You can [read more about lists](https://htmhell.dev/adventcalendar/2023/13/#lessulgreater-and-lessolgreater) if you're curious.

Coincidentally, the `<ul>` is also the reason we can't attach a shadow-tree (see [list of allowed elements](https://developer.mozilla.org/en-US/docs/Web/API/Element/attachShadow#elements_you_can_attach_a_shadow_to)). Changing `<ul>`+`<li>` to `<div>`s would solve this issue as well.

Reply:
Okay, change the elements to `<div>`s and wrapped the inner div in a `ShadowRoot` in this [commit](https://github.com/iTwin/iTwinUI/pull/2091/commits/8629bd125abce0a4a10b0826f88d2f8ee4c0098c).